src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -6483,41 +6514,71 @@ public void Invoke(Task completingTask)
         /// <exception cref="ArgumentException">
         /// The <paramref name="tasks"/> collection contained a null task, or was empty.
         /// </exception>
-        public static Task<Task> WhenAny(IEnumerable<Task> tasks)
+        public static Task<Task> WhenAny(IEnumerable<Task> tasks) =>
+            WhenAny<Task>(tasks);
+
+        /// <summary>
+        /// Creates a task that will complete when any of the supplied tasks have completed.
+        /// </summary>
+        /// <param name="tasks">The tasks to wait on for completion.</param>
+        /// <returns>A task that represents the completion of one of the supplied tasks.  The return Task's Result is the task that completed.</returns>
+        /// <remarks>
+        /// The returned task will complete when any of the supplied tasks has completed.  The returned task will always end in the RanToCompletion state
+        /// with its Result set to the first task to complete.  This is true even if the first task to complete ended in the Canceled or Faulted state.
+        /// </remarks>
+        /// <exception cref="ArgumentException">
+        /// The <paramref name="tasks"/> collection contained a null task, or was empty.
+        /// </exception>
+        private static Task<TTask> WhenAny<TTask>(IEnumerable<TTask> tasks) where TTask : Task
         {
             // Skip a List allocation/copy if tasks is a collection
-            if (tasks is ICollection<Task> taskCollection)
+            if (tasks is ICollection<TTask> tasksAsCollection)
             {
-                // Take a more efficient path if tasks is actually an array
-                if (tasks is Task[] taskArray)
+                // Take a more efficient path if tasks is actually a list or an array. Arrays are a bit less common,
+                // since if argument was strongly-typed as an array, it would have bound to the array-based overload.
+                if (tasks is List<TTask> tasksAsList)
                 {
-                    return WhenAny(taskArray);
+                    return WhenAny((ReadOnlySpan<TTask>)CollectionsMarshal.AsSpan(tasksAsList));
+                }
+                if (tasks is TTask[] tasksAsArray)
+                {
+                    return WhenAny((ReadOnlySpan<TTask>)tasksAsArray);
```

Comment:
We need the `ReadOnlySpan<TTask>` overload to be able to support both arrays and lists.  If we also had a `TTask[]` overload, it would either need to duplicate exactly what the span overload does, or just delegate to the span-based overload, at which point there's no benefit to having the `TTask[]` overload, because these call sites can just go directly to the span-based one.